Add the feature of search images from giphy

shuowpro opened the pull request:
* Add the component ImageFlow to show the list of images
* Add the Feature of searching images
* Debounce the searching process

Commit:
add the feature of search images from giphy

IOBveritas commented on src/App.js:
```diff
@@ -1,25 +1,42 @@
 import './App.css';
 
 class App extends Component {
+  constructor(props) {
```
```js
class App extends Component {
  state = { 
    searchText: '',
    imageAddrs: []
  } 
  timeout = null
...
```
What do you think about such refactoring? Does it have any advantages over `contructor` approach?

shuowpro replied:
This is a syntax sugar. The advantage of using this is that we can make the code to be clearer and we do not need to worry about forget 
```js
constructor(props) {
  super(props) // this line
}
```

IOBveritas commented on src/App.js:
```diff
@@ -1,25 +1,42 @@
+    this.setState({
+      searchText: event.target.value,
+    });
+    clearTimeout(this.timeout);
```
Can we rewrite this function using `async`/`await` and get rid of `setTimeout`, `clearTimeout`?

shuowpro replied:
Yes, I have rewritten it and submitted it within the next commit.

IOBveritas replied:
Thank you!

IOBveritas commented on src/App.js:
```diff
@@ -1,25 +1,42 @@
+    clearTimeout(this.timeout);
+    this.timeout = setTimeout(searchText => {
+      axios.get(`https://api.giphy.com/v1/gifs/search?api_key=YWvK8etOQgpDuzevIRng2Np91aSn38PF&q=${searchText}`)
+      .then(res => res.data.data)
```
I propose to rewrite API call like this:
```js
axios.get(`https://api.giphy.com/v1/gifs/search?api_key=YWvK8etOQgpDuzevIRng2Np91aSn38PF&q=${searchText}`)
  .then(({ data }) => { 
    const addrs = data.data.map(({ images }) => images.original.url)
    this.setState({ imageAddrs: addrs })
  })
```
Do you think that using `then` multiple times for synchronous operations can have any potential problems?

shuowpro replied:
Yes, because the `then` is asynchronous, if we split the synchronous operation apart with `then`, we will get some of the unexpected behavior because we just make the synchronous operation to executed asynchronously.

IOBveritas commented on src/ImageFlow.jsx:
```diff
@@ -0,0 +1,27 @@
+  imageAddrs: [],
+}
+
+export default ImageFlow;
```
What are potential downsides of default exports compared to named exports in your opinion? e.g.:
```
export const ImageFlow = props => {         // named export
...
export default ImageFlow                    // default export
```

shuowpro replied:
The default exports do not export any name, so that the intellisense doesn't work. Every people can rename the modules, so that the refactoring will become a nightmare.

IOBveritas commented on src/App.js:
```diff
@@ -1,25 +1,42 @@
+    this.timeout = null;
+  }
+
+  handleUpdateSearchText = event => {
```
Why do you use arrow function here instead of a regular class method?

shuowpro replied:
The reason for us to use the arrow function is that the arrow function will automatically bind the `this` to the current class. If we do not use the arrow function. We will need to use the following code like this

```js
constructor(props) {
  super(props);
  this.handleUpdateSearchText = this.handleUpdateSearchText.bind(this);
}
```

IOBveritas reviewed (commented):
Hi Shuo! :wave:
Let's have a small discussion about giphy app.
Please, can you answer questions in Pull Requests and also these 2 questions: 
1) How testable is this code? If this is your sprint task, what tests would you write? Would you make any changes to make the code more testable?
2) How could this App can be improved? give rein to your imagination here :unicorn: :)

Commit:
refactory the fetch images with await and simplify the code

shuowpro commented:
Hello @IOBveritas !

I have updated the code based on your suggestion, please feel free to check it

I will answer the 2 questions as below.

1. This project contains 2 main components: one is stateful and another is stateless. The stateless component is more testable. we can pass with an array of string, an empty array, or not pass any data. The stateful component is more complex. We can use the enzyme to mock the behavior of input message and test if the `handleFetchImages` is called, I can also mock the http request to test the `handleFetchImages`. I also need to test if the components are successfully rendered. In my latest code. I have split the `handleUpdateSearchText` and `handleFetchImages` into 2 separate code. each one is more easily to be tested. We can also use the selenium or other tools to test the functionality and the styles.
2. There are a lot of thing that we can do to improve. I have not put any css into this project. I think we can style it in the future to make it prettier. One of the best design here is to use the `waterfall`. We have not use any state management library in this project. We can just introduce the [redux](https://github.com/reduxjs/redux) or [mobx](https://github.com/mobxjs/mobx) if our project become more and more complicated in the future. The ajax library we used here is not easy for us to use in order to combine it with [redux](https://github.com/reduxjs/redux). We can use the [redux-observable](https://github.com/redux-observable/redux-observable) or [redux-saga](https://github.com/redux-saga/redux-saga) to replace the [axios](https://github.com/axios/axios) that we are currently used.
